Panic on autopeering: index out of range

Running docker-network and docker tests encounters a panic on random nodes:
```
2024-02-27T08:24:54Z	DEBUG  	P2P.Autopeering                     	Too many autopeering peers connected 1, disconnecting some 
panic: runtime error: index out of range [-1]

goroutine 164 [running]:
github.com/iotaledger/iota-core/pkg/network/p2p/autopeering.(*Manager).discoverAndDialPeers(0xc0004d0e70)
	/scratch/iota-core/pkg/network/p2p/autopeering/autopeering.go:197 +0x307
github.com/iotaledger/iota-core/pkg/network/p2p/autopeering.(*Manager).discoveryLoop(0xc0004d0e70)
	/scratch/iota-core/pkg/network/p2p/autopeering/autopeering.go:179 +0x94
created by github.com/iotaledger/iota-core/pkg/network/p2p/autopeering.(*Manager).Start.func1 in goroutine 552
	/scratch/iota-core/pkg/network/p2p/autopeering/autopeering.go:102 +0x75c
```

To see the panic, start docker-network and wait for a while.

--- a/pkg/network/p2p/autopeering/autopeering.go
+++ b/pkg/network/p2p/autopeering/autopeering.go
@@ -3,6 +3,7 @@ package autopeering
 import (
 	"context"
 	"fmt"
+	"math/rand"
 	"sync"
 	"sync/atomic"
 	"time"
@@ -183,19 +184,38 @@ func (m *Manager) discoveryLoop() {
 	}
 }
 
+func randomSubset[T any](slice []T, n int) []T {
+	if n >= len(slice) {
+		rand.Shuffle(len(slice), func(i, j int) {
+			slice[i], slice[j] = slice[j], slice[i]
+		})
+
+		return slice
+	}
+
+	subset := make([]T, n)
+	indices := rand.Perm(len(slice)) // Get a slice of random unique indices
+	for i := 0; i < n; i++ {
+		subset[i] = slice[indices[i]]
+	}
+
+	return subset
+}
+
 func (m *Manager) discoverAndDialPeers() {
 	autopeeringNeighbors := m.networkManager.AutopeeringNeighbors()
 	peersToFind := m.maxPeers - len(autopeeringNeighbors)
 	if peersToFind == 0 {
-		m.logger.LogDebugf("%d autopeering peers connected, not discovering new ones. (max %d)", len(autopeeringNeighbors), m.maxPeers)
+		m.logger.LogDebugf("%d autopeering neighbors connected, not discovering new ones. (max %d)", len(autopeeringNeighbors), m.maxPeers)
 		return
 	}
 
 	if peersToFind < 0 {
-		m.logger.LogDebugf("Too many autopeering peers connected %d, disconnecting some", -peersToFind)
-		for i := peersToFind; i < 0; i++ {
-			if err := m.networkManager.DropNeighbor(autopeeringNeighbors[i].Peer().ID); err != nil {
-				m.logger.LogDebugf("Failed to disconnect neighbor %s", autopeeringNeighbors[i].Peer().ID)
+		neighborsToDrop := randomSubset(autopeeringNeighbors, -peersToFind)
+		m.logger.LogDebugf("Too many autopeering neighbors connected %d, disconnecting some", len(neighborsToDrop))
+		for _, peer := range neighborsToDrop {
+			if err := m.networkManager.DropNeighbor(peer.Peer().ID); err != nil {
+				m.logger.LogDebugf("Failed to disconnect neighbor %s", peer.Peer().ID)
 			}
 		}
 
@@ -205,7 +225,7 @@ func (m *Manager) discoverAndDialPeers() {
 	findCtx, cancel := context.WithTimeout(m.ctx, 10*time.Second)
 	defer cancel()
 
-	m.logger.LogDebugf("%d autopeering peers connected. Discovering new peers for namespace %s", len(autopeeringNeighbors), m.namespace)
+	m.logger.LogDebugf("%d autopeering neighbors connected. Discovering new peers for namespace %s", len(autopeeringNeighbors), m.namespace)
 	peerChan, err := m.routingDiscovery.FindPeers(findCtx, m.namespace)
 	if err != nil {
 		m.logger.LogWarnf("Failed to find peers: %s", err)

